File: generators/kotlin_ktor_client_generator/test/src/jvmTest/kotlin/com/example/kitchensink/client/FormBodiesTest.kt
package com.example.kitchensink.client

import com.example.kitchensink.client.support.buildTestClient
import io.ktor.client.engine.mock.respond
import io.ktor.client.request.HttpRequestData
import io.ktor.http.HttpMethod
import io.ktor.http.HttpStatusCode
import io.ktor.http.content.OutgoingContent
import io.ktor.http.parseQueryString
import kotlinx.coroutines.test.runTest
import kotlin.test.Test
import kotlin.test.assertContentEquals
import kotlin.test.assertEquals
import kotlin.test.assertTrue

// Exercises the four non-default-JSON request-body encodings this generator supports (see
// lib/operations.js's pickBodyContent/buildRequestBody and api_client.kt.j2's FormDataContent/
// MultiPartFormDataContent/ContentType.parse branches) against a MockEngine - same style as
// PetsApiTest.
class FormBodiesTest {
    @Test
    fun `subscribeToPet sends the body as application_x-www-form-urlencoded`() = runTest {
        var captured: HttpRequestData? = null
        val client = buildTestClient { request ->
            captured = request
            respond("", HttpStatusCode.NoContent)
        }
        val api = PetsApi(client, "https://example.test")

        api.subscribeToPet(petId = "1", body = PetSubscription(email = "me@example.com", notify = true))

        assertEquals(HttpMethod.Post, captured!!.method)
        val body = captured!!.body as OutgoingContent.ByteArrayContent
        assertTrue(body.contentType?.toString()?.startsWith("application/x-www-form-urlencoded") == true)
        val params = parseQueryString(String(body.bytes()))
        assertEquals("me@example.com", params["email"])
        assertEquals("true", params["notify"])
    }

    @Test
    fun `uploadPetPhoto sends the body as multipart_form-data`() = runTest {
        var captured: HttpRequestData? = null
        val client = buildTestClient { request ->
            captured = request
            respond("", HttpStatusCode.NoContent)
        }
        val api = PetsApi(client, "https://example.test")

        api.uploadPetPhoto(petId = "1", body = PetPhotoUpload(caption = "cute", photo = "raw-bytes"))

        assertEquals(HttpMethod.Post, captured!!.method)
        // MultiPartFormDataContent is streamed (WriteChannelContent), not a simple byte buffer -
        // asserting on the auto-generated boundary'd Content-Type is enough to prove the generated
        // code picked the multipart builder and didn't crash building/sending it.
        assertTrue(captured!!.body.contentType?.toString()?.startsWith("multipart/form-data") == true)
    }

    @Test
    fun `setPetNotes sends and parses a text_plain body as a plain String`() = runTest {
        var captured: HttpRequestData? = null
        val client = buildTestClient { request ->
            captured = request
            respond("echo: " + String((request.body as OutgoingContent.ByteArrayContent).bytes()), HttpStatusCode.OK)
        }
        val api = PetsApi(client, "https://example.test")

        val result = api.setPetNotes(petId = "1", body = "likes belly rubs")

        assertEquals(HttpMethod.Post, captured!!.method)
        // Ktor's setBody(String) wraps a raw string into TextContent, which appends "; charset=
        // UTF-8" to whatever Content-Type was set for a text/* type - not something the generated
        // code controls, so this only checks the media type prefix, same as the multipart/
        // urlencoded assertions above.
        assertTrue(captured!!.body.contentType?.toString()?.startsWith("text/plain") == true)
        assertEquals("likes belly rubs", String((captured!!.body as OutgoingContent.ByteArrayContent).bytes()))
        assertEquals("echo: likes belly rubs", result)
    }

    @Test
    fun `uploadPetAvatar sends and parses an application_octet-stream body as raw bytes`() = runTest {
        val bytes = byteArrayOf(0x89.toByte(), 0x50, 0x4e, 0x47)
        var captured: HttpRequestData? = null
        val client = buildTestClient { request ->
            captured = request
            respond(bytes, HttpStatusCode.OK)
        }
        val api = PetsApi(client, "https://example.test")

        val result = api.uploadPetAvatar(petId = "1", body = bytes)

        assertEquals(HttpMethod.Put, captured!!.method)
        assertEquals("application/octet-stream", captured!!.body.contentType?.toString())
        assertContentEquals(bytes, (captured!!.body as OutgoingContent.ByteArrayContent).bytes())
        assertContentEquals(bytes, result)
    }
}

File: generators/typescript_fetch_client_generator/test/src/formBodies.test.ts
import { test } from "node:test";
import assert from "node:assert/strict";
import { PetsClient } from "../generated/apis/PetsClient.js";
import { createFetchStub } from "./support/fetchStub.js";

test("subscribeToPet sends the body as application/x-www-form-urlencoded", async () => {
  const { fetch, calls } = createFetchStub(() => ({ status: 204 }));
  await new PetsClient({ baseUrl: "https://example.test", fetch }).subscribeToPet("1", {
    body: { email: "me@example.com", notify: true },
  });
  assert.equal(calls[0]!.headers["content-type"], "application/x-www-form-urlencoded");
  const params = new URLSearchParams(calls[0]!.body as string);
  assert.equal(params.get("email"), "me@example.com");
  assert.equal(params.get("notify"), "true");
});

test("uploadPetPhoto sends the body as a native FormData with no Content-Type set", async () => {
  const { fetch, calls } = createFetchStub(() => ({ status: 204 }));
  const photo = new Blob(["raw-bytes"], { type: "image/jpeg" });
  await new PetsClient({ baseUrl: "https://example.test", fetch }).uploadPetPhoto("1", {
    body: { caption: "cute", photo },
  });
  // fetch itself sets Content-Type (with the multipart boundary) once it sees a FormData body -
  // the stub never applies that step, so its absence here just proves the generated code didn't
  // set one itself (which would omit the boundary and break a real request).
  assert.equal(calls[0]!.headers["content-type"], undefined);
  assert.ok(calls[0]!.body instanceof FormData);
  const body = calls[0]!.body as FormData;
  assert.equal(body.get("caption"), "cute");
  // FormData.append wraps a Blob value into a File (Node's WHATWG FormData implementation, same
  // as browsers) - not the same object anymore, so compare content/type instead of reference.
  const uploadedPhoto = body.get("photo") as Blob;
  assert.ok(uploadedPhoto instanceof Blob);
  assert.equal(uploadedPhoto.type, "image/jpeg");
  assert.equal(uploadedPhoto.size, photo.size);
});

test("setPetNotes sends and parses a text/plain body as a plain string", async () => {
  const { fetch, calls } = createFetchStub((req) => ({ status: 200, body: `echo: ${req.body as string}` }));
  const result = await new PetsClient({ baseUrl: "https://example.test", fetch }).setPetNotes("1", {
    body: "likes belly rubs",
  });
  assert.equal(calls[0]!.headers["content-type"], "text/plain");
  assert.equal(calls[0]!.body, "likes belly rubs");
  assert.equal(result, "echo: likes belly rubs");
});

test("uploadPetAvatar sends and parses an application/octet-stream body as raw bytes", async () => {
  const bytes = new Uint8Array([0x89, 0x50, 0x4e, 0x47]);
  const { fetch, calls } = createFetchStub(() => ({ status: 200, body: bytes }));
  const result = await new PetsClient({ baseUrl: "https://example.test", fetch }).uploadPetAvatar("1", {
    body: bytes,
  });
  assert.equal(calls[0]!.headers["content-type"], "application/octet-stream");
  assert.ok(calls[0]!.body instanceof Uint8Array);
  assert.deepEqual(calls[0]!.body, bytes);
  assert.ok(result instanceof Uint8Array);
  assert.deepEqual(result, bytes);
});
